### ISSUE TITLE
Assertion at..., condition '<disabled>' not met.

### ISSUE THREAD

**Issue by CincySteve** [dotnet/runtime#42628]
I just published to Azure a Blazor Wasm SPA I am developing with VS Community 2019.  The app starts up successfully in Chrome via the VS debugger, but the published version triggers the exception below (also in Chrome).  The exception message, _ gives me no clue as to the problem_.  This looks similar to this [previous issue dotnet/aspnetcore#16818](https://github.com/mono/mono/issues/16818) that has been fixed.  I published 2-3 months ago a much simpler early version of the app, which functionally did what's described below, and it worked at that time.  

Curiously, after I first published a few minutes ago, my console log messages indicated the app got through almost all of my main layout's OnInitializedAsync code, including a JSRuntime.InvokeAsync to my JavaScipt initialization code that opens an IndexedDB and returns its Index of keys.  But when I tried re-publishing and got the same exception, none of my initialization code ran according to the console log.  I can post my OnInitializedAsync code, but since the exception is now happening before any of this code executes, I'm not sure it will be helpful.  Any idea what's going on?

Thanks.  Steve

![App Startup Exception](https://user-images.githubusercontent.com/18602722/93339179-441ead80-f7f9-11ea-9720-a27c236dfd58.JPG)

**Comment by mkArtakMSFT** [comment]
@eerhardt this seems to be caused by Linker

**Comment by eerhardt** [comment]
Looks like this assert is failing:

https://github.com/dotnet/runtime/blob/dbc2d36be93604d9820eefa35e6e234aecf6ff24/src/mono/mono/metadata/threads.c#L354

cc @lewing

**Comment by lewing** [comment]
@CincySteve do you have a test case we can use to reproduce this?

**Comment by lewing** [comment]
Does this fail always or only when debugging?

**Comment by lewing** [comment]
Does this still fail if you add

```
    <PublishTrimmed>false</PublishTrimmed> 
```

To your project file?

**Comment by CincySteve** [comment]
Unfortunately, the only app I have that experiences the problem is a fairly big one: 40+ components, 40+ significant c# classes, etc.

**Comment by CincySteve** [comment]
@lewing I'm afraid I don't know what adding PublishTrimmed to my "project file" means.  I tried to add it in several places, but it was not a recognizable component.  Which specific project file do you mean?

**Comment by lewing** [comment]
@CincySteve in the blazor wasm specific csproj inside a property group.

**Comment by lewing** [comment]
@lertoo @modulardev2 if either of you have a reproduction you can share it would be very helpful.

**Comment by BrzVlad** [comment]
@CincySteve While there is always the chance that something goes unexpectedly wrong somewhere, based on the stacktrace that you provided this is not a bug but rather a current limitation of the interpreter. The interpreter will fail to compile methods that are extremely large (it requires more than 32kb of local space) and this is the first situation where I hear of this happening. Fixing this is quite a bit of effort, so it wouldn't make it to 5.0. If you are over on the DotNetEvolution discord server, we can discuss it there in private and I can guide you to find which method fails to compile and confirm if this limitation is actually hit.

**Comment by vargaz** [comment]
Maybe we should error out instead of just asserting to make it easier to find the offending method.
The JIT throws an InvalidProgram exception for some extremely large methods with the 
message 'Method <foo> is too complex.".

**Comment by CincySteve** [comment]
@BrzVlad Thanks for the help.  I just restored a version of my app from 2 days ago.  It does not trigger the exception.  Based on a folder comparison, it looks like I made a bunch of changes, so I would appreciate your private help in DotNetEvolution to find the one (or more) that hit the limit.  So I and others are clear, the problem is with a METHOD that is too big, not something else (like a class).  I am registered in DotNetEvolution and have sent you a message.  Thanks again.

**Comment by modulardev2** [comment]
> @lertoo @modulardev2 if either of you have a reproduction you can share it would be very helpful.

Sure thing. I created a project here: https://github.com/modulardev2/ThreadException

It's simply the standard blazor web assembly/hosted project template, with the FetchData OnInitializedAsync changed to:
``
protected override async Task OnInitializedAsync()
    {
        Console.WriteLine(await Http.GetStringAsync("WeatherForecast"));
    }
``

It works when I run it locally, but as you can see if you go to the live site it doesn't work: https://pws.northcentralus.cloudapp.azure.com:49190/fetchdata

If it helps, the site is running on IIS in Azure

**Comment by BrzVlad** [comment]
@modulardev2 Thanks. Can't reproduce from VS, but it fails when it is published to a local IIS instance.

**Comment by medeirosraul** [comment]
Same error here, just only in published project.

**Comment by eerhardt** [comment]
Setting this as `5.0` and `blocking` as this is a pretty important scenario that is broken in Blazor WASM. I can reproduce it with the following steps:

```
dotnet new blazorwasm

edit the Pages/FetchData.razor file to add 
    Console.WriteLine(await Http.GetStringAsync("/"));
inside of the OnInitializedAsync method

dotnet publish -c Release

dotnet serve -d .\bin\Release\net5.0\publish\wwwroot\

navigate to http://localhost:8080/ and click on the "Fetch data" tab
```

(note you need to install https://github.com/natemcmaster/dotnet-serve to use dotnet serve)

I also verified that this worked on `5.0-preview8`, but fails on `5.0-rc1`.

**Comment by medeirosraul** [comment]
In my case, error occurs also in DEBUG published in Azure Webapps. So I think this is not a Release mode particularity.

**Comment by eerhardt** [comment]
> So I think this is not a Release mode particularity.

Agreed. It repros with `dotnet publish -c Debug` as well. Force of habit that I always `publish -c Release`. 😉

**Comment by CincySteve** [comment]
Just so the discussion doesn't get too narrow, as the originator of this issue, I would like to note that I have experienced the assertion exception in both the debugger under VS and with the same app published to Azure.  And, my app does not invoke Http.GetStringAsync as some who have had a similar or the same exception have.  

I have tried to replicate it under .Net 5.0 RC1, but think I didn't do that properly.  In any case, I will start tomorrow to incrementally change my app from the last version that worked to one that does not by comparing the changes I made between the two versions.  I hope that will help.

**Comment by BrzVlad** [comment]
@CincySteve I suggest opening a separate issue together with any new findings, since this one was hijacked by a separate .net 5 rc1 regression.

**Comment by SamMonoRT** [comment]
@lewing @mkArtakMSFT I believe we can move the Blocking-release label and 5.0 Milestone from here to the linked bug.

**Comment by lewing** [comment]
@SamMonoRT mono/linker doesn't have the labels, I'm not sure how it is usually handled?

This is blocking on https://github.com/mono/linker/issues/1507

**Comment by SamMonoRT** [comment]
@eerhardt - Leaving this open as a Blocked issue - Not an interpreter issue, update area label to area-Meta

**Comment by CincySteve** [comment]
After a lot of digging and help from @BrzVlad , I have found that he was correct in diagnosing my problem (not necessarily the others) as a method that is too large.  I load some data lists in Program.cs via calls to static class methods, one of which is very large.  These classes are generated by a custom utility I wrote using external data sources.  I'm not sure why this started to fail only recently (this code has been running for at least 2 months, which is why I didn't think about it).  And of course, the Assertion exception is a mysterious symptom.  Anyway, thanks once again to @BrzVlad.

**Comment by eerhardt** [comment]
Update - the underlying [linker issue](https://github.com/mono/linker/issues/1507) has been fixed and is being [backported to 5.0](https://github.com/mono/linker/pull/1513). Once that is merged we should be able to close this issue.

**Comment by Anipik** [comment]
@eerhardt  is there any work required on the runtime side for this issue for rc2 or GA ?

**Comment by eerhardt** [comment]
I don't believe so. To fix this we will need:

1. The above linker change to be ported to the 5.0 branch.
2. The new 5.0 linker to be merged into the dotnet/sdk 5.0 branch.

**Comment by marek-safar** [comment]
The changes were backported to 5.0 GA branch and merged to dotnet/sdk

**Comment by truthz03** [comment]
Hi, I have updated to RC2 today but the error still is present.

The problem seems to be this line from a Nuget Package I'm using:
await Http.GetJsonAsync<TableItem>(ApiUrl + SearchTermParam(SearchTerm));

Is this Bug not fixed unit now?

**Comment by lewing** [comment]
@truthz03 RC2 was frozen before the fix went in.  The fix will be part of 5.0 final.

**Comment by truthz03** [comment]
Ok. Thanks. Than I will wait for it

**Comment by joer33304** [comment]
This issue is NOT fixed. I upgrade to the latest 5 net 5 sdk. Works locally, error appears when deploying to azure on windows hosted app service.

**Comment by joer33304** [comment]
Just for others coming here, "<PublishTrimmed>false</PublishTrimmed> " works for me to get rid of the exception

**Comment by lewing** [comment]
@joer33304 can you please open a new issue with the details of the problem you are encountering?
